forem/forem
# Add logged-in vs. logged-out targeting option for Display Ads #18390

## What type of PR is this? (check all applicable)


## Description
Adding the ability to configure a display ad to show to logged-in, logged-out or all users.

Loom recording to test Display Ad's:
https://www.loom.com/share/d30081d5810e4079a0df59dcc14da276

Note: I haven't thought about tracking as yet especially for logged out users.

## Related Tickets & Documents



- Related Issue #https://github.com/forem/forem/issues/18390

## QA Instructions, Screenshots, Recordings

Some admin screenshots:

<img width="1510" alt="Screenshot 2022-09-01 at 14 07 09" src="https://user-images.githubusercontent.com/2786819/187910143-a03fa2e3-51cb-493e-a49b-977aea636a6b.png">

<img width="1512" alt="Screenshot 2022-09-01 at 14 06 52" src="https://user-images.githubusercontent.com/2786819/187910123-7adea553-2694-4212-bc28-544074378fec.png">

Loom recording to test Display Ad's:
https://www.loom.com/share/d30081d5810e4079a0df59dcc14da276

Display Ads are a little complex to test on the UI because we randomly select the display ad from a pool of display ads that meet the correct conditions. Hence you will see 1 of X amount of display ads that may or may not change on reload. 

The most ideal way to test this can be: 
- First clear all you display ads using the console `DisplayAd.delete_all`
- Then, go to http://localhost:3000/admin/customization/display_ads/ and add a new Display Ad. You want your ads to be published and approved and set the placement  area to sidebar_right. 
- For the first ad, set the `Display to user group` field to logged in for this one. Thereafter, create another Display Ad and select logged_out for this one. 
- Now, login and logout and you should see the appropriate ad render on the right hand side sidebar on the home page. 

- To test "all", add another display ad and set the `Display to user group` field to "All Users" on this one. 
- Now go back to the home page and refresh a couple of times, you should see it appear on both logged in and logged out. Remember, the ad is getting randomly selected and so you may not see it after a couple of refreshes. In order to increase the odds you can unpublish and unapprove the other two ads (i.e. the logged in and the logged out one).

### UI accessibility concerns?
None, very minimal UI changes.

## Added/updated tests?

      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Storybook](https://storybook.forem.com/) (for Crayons components)
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/hpI0G7dgp87Px27IJy/giphy.gif)
